src/filter.rs

```diff
@@ -0,0 +1,322 @@
+}
+
+extern fn stream_close(stream: *mut git_writestream) -> c_int {
+    let ws = stream as *mut Stream;
+    panic::wrap(|| unsafe {
+        match (*ws).handler.close() {
+            Ok(..) => 0,
+            Err(e) => e.raw_code() as c_int,
+        }
+    }).unwrap_or(-1)
+}
+
+extern fn filter_init(filter: *mut git_filter) -> c_int {
+    let filter = filter as *mut Filter;
+    panic::wrap(|| unsafe {
+        match (*filter).instance.init() {
+            Ok(..) => 0,
+            Err(e) => e.raw_code() as c_int,
+        }
+    }).unwrap_or(-1)
+}
+
+extern fn filter_shutdown(filter: *mut git_filter) {
+    let _ = unsafe { Box::from_raw(filter as *mut Filter) };
+}
+
+extern fn filter_stream_init(out: *mut *mut raw::git_writestream,
+                             filter: *mut git_filter,
+                             _payload: *mut *mut c_void,
+                             src: *const raw::git_filter_source,
+                             next: *mut raw::git_writestream) -> c_int
+{
+    let filter = filter as *mut Filter;
+    let next = RawWriteStream { stream: next };
+    let source = FilterSource::from_raw(src);
+    panic::wrap(|| unsafe {
+        match (*filter).instance.open_stream(source, next) {
+            Ok(s) => {
+                let mut stream = Box::new(Stream {
+                    parent: git_writestream {
+                        write: stream_write,
+                        close: stream_close,
+                        free: stream_free,
+                    },
+                    handler: s,
+                });
+                let p = (&mut *stream) as *mut Stream as *mut git_writestream;
+                *out = p;
+                mem::forget(stream);
+                0
+            },
+            Err(e) => e.raw_code() as c_int,
+        }
+    }).unwrap_or(-1)
+}
+
+impl Filter {
+    /// create a command filter
+    fn new<T: StreamFilter>(user_filter: T, attrs: Option<&str>) -> Filter {
+        let mut f = Filter {
+            filter: git_filter {
+                version: raw::GIT_FILTER_VERSION,
+                attributes: ptr::null(),
+                initialize: Some(filter_init),
+                shutdown: Some(filter_shutdown),
+                check: None,
+
+                // do not use legency filter model
+                apply: None,
+                stream: Some(filter_stream_init),
+
+                // stream filter do not need this
+                cleanup: None,
+            },
+            attrs: attrs.map(|e| { CString::new(e).unwrap() }),
+            instance: Box::new(user_filter),
+        };
+        if let Some(v) = f.attrs.as_ref() {
+            f.filter.attributes = v.as_ptr();
+        }
+        f
+    }
+}
+
+/// register a command filter.
+pub fn register<T: StreamFilter>(name: &str, attributes: Option<&str>,
+                                 priority: i32, filter: T)
+                                 -> Result<(), Error> {
+    let f = Filter::new(filter, attributes);
```

Comment:
Could this call be pushed into `_register`? The `filter` is just boxed up in there anyway, right?